drivers/adc/adc_ads1119.c

```diff
@@ -0,0 +1,522 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ * Copyright (c) 2020 Innoseis BV
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/adc.h>
+#include <logging/log.h>
+#include <drivers/i2c.h>
+#include <zephyr.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER 1
+#include "adc_context.h"
+
+#define DT_DRV_COMPAT ti_ads1119
+
+LOG_MODULE_REGISTER(ADS1119, CONFIG_ADC_LOG_LEVEL);
+
+#define ADS1119_GEN_MASK(shift, size)  (BIT_MASK(size) << shift)
+
+#define ADS1119_CONFIG_SHIFT_VREF 0
+#define ADS1119_CONFIG_SIZE_VREF 1
+#define ADS1119_CONFIG_MASK_VREF ADS1119_GEN_MASK(ADS1119_CONFIG_SHIFT_VREF, \
+						  ADS1119_CONFIG_SIZE_VREF)
+#define ADS1119_CONFIG_SHIFT_CM 1
+#define ADS1119_CONFIG_SIZE_CM 1
+#define ADS1119_CONFIG_MASK_CM ADS1119_GEN_MASK(ADS1119_CONFIG_SHIFT_CM, \
+						ADS1119_CONFIG_SIZE_CM)
+#define ADS1119_CONFIG_SHIFT_DR 2
+#define ADS1119_CONFIG_SIZE_DR 2
+#define ADS1119_CONFIG_MASK_DR ADS1119_GEN_MASK(ADS1119_CONFIG_SHIFT_DR, \
+						ADS1119_CONFIG_SIZE_DR)
+#define ADS1119_CONFIG_SHIFT_GAIN 4
+#define ADS1119_CONFIG_SIZE_GAIN 1
+#define ADS1119_CONFIG_MASK_GAIN ADS1119_GEN_MASK(ADS1119_CONFIG_SHIFT_GAIN, \
+						  ADS1119_CONFIG_SIZE_GAIN)
+#define ADS1119_CONFIG_SHIFT_MUX 5
+#define ADS1119_CONFIG_SIZE_MUX 3
+#define ADS1119_CONFIG_MASK_MUX ADS1119_GEN_MASK(ADS1119_CONFIG_SHIFT_MUX, \
+						 ADS1119_CONFIG_SIZE_MUX)
+
+#define ADS1119_CONFIG(flag, n) (ADS1119_CONFIG_MASK_##flag & (n << ADS1119_CONFIG_SHIFT_##flag))
+
+#define ADS1119_STATUS_SHIFT_ID 0
+#define ADS1119_STATUS_SIZE_ID 7
+#define ADS1119_CONFIG_MASK_ID ADS1119_GEN_MASK(ADS1119_STATUS_SHIFT_ID, \
+						ADS1119_STATUS_SIZE_ID)
+
+#define ADS1119_STATUS_SHIFT_READY 7
+#define ADS1119_STATUS_SIZE_READY 1
+#define ADS1119_STATUS_MASK_READY ADS1119_GEN_MASK(ADS1119_STATUS_SHIFT_READY, \
+						   ADS1119_STATUS_SIZE_READY)
+
+#define ADS1119_REG_SHIFT 2
+
+#define ADS1119_RESOLUTION 15
```

Comment:
See discussion here about the 16 bit resolution definition for differential and single-ended
https://github.com/zephyrproject-rtos/zephyr/pull/40477#discussion_r765875293

How I had to implement this was to cache if it was configured for a differential and then validate the configured resolution.
https://github.com/zephyrproject-rtos/zephyr/blob/22fd253fead279aa3c33a5dda51ada410a11782d/drivers/adc/adc_ads1x1x.c#L450